static/scripts/rewards/web3/connect-wallet.ts

```diff
@@ -18,18 +22,148 @@ export async function connectWallet(): Promise<JsonRpcSigner | null> {
+
+  const latencies = handler.getLatencies();
+  const latenciesArray = Object.entries(latencies).map(([url, latency]) => ({ url, latency }) as { url: string; latency: number });
+  const sorted = latenciesArray.sort((a, b) => a.latency - b.latency);
+
+  let toSuggest = sorted[0];
+
+  let isOkay = false;
+
+  for await (const { url } of sorted) {
+    const _url = url.split("__")[1];
+    if (_url !== appUrl) {
+      provider = new ethers.providers.JsonRpcProvider(_url);
+
+      isOkay = await stressTestWalletRpc(provider);
+
+      if (isOkay) {
+        toSuggest = { url: _url, latency: latencies[url] };
+        break;
+      }
+    }
+  }
+
+  if (!isOkay) {
+    toaster.create("error", "We failed to find a more reliable RPC for you. Please try again later if you have network issues.");
+    return;
+  }
+
+  try {
+    await addHandlerSuggested(wallet, toSuggest.url);
+  } catch (error) {
+    toaster.create("info", `${toSuggest.url}`, Infinity);
+  }
+}
+
+async function addHandlerSuggested(provider: ethers.providers.Web3Provider, url: string) {
+  const symbol = app.networkId === 1 ? "ETH" : "XDAI";
+  const altSymbol = app.networkId === 1 ? "eth" : "xdai";
+  const altSymbol2 = app.networkId === 1 ? "Eth" : "xDai";
+
+  if (mediaQuery.matches) {
+    /**
+     * https://github.com/MetaMask/metamask-mobile/issues/9519
+     *
+     * Until this is resolved it is not possible for us to add a network on mobile
+     * so we'll show a toast suggesting they do it manually
+     */
+
+    toaster.create("info", `${url}`, Infinity);
+    return;
+  }
+
+  // It will not work unless the symbols match, so we try them all
+  for (const _symbol of [symbol, altSymbol, altSymbol2]) {
+    // this does not work on mobile yet
+    await addProvider(provider, url, _symbol, app.networkId);
+  }
+}
+
+async function addProvider(provider: ethers.providers.Web3Provider, url: string, symbol: string, chainId: number | null) {
+  const _chainId = chainId || (await provider.getNetwork()).chainId;
+  try {
+    await provider.send("wallet_addEthereumChain", [
+      {
+        chainId: `0x${_chainId.toString(16)}`,
+        chainName: _chainId === 1 ? "Ethereum" : "Gnosis",
+        nativeCurrency: {
+          name: _chainId === 1 ? "ETH" : "XDAI",
+          symbol,
+          decimals: 18,
+        },
+        rpcUrls: [url],
+        blockExplorerUrls: [`https://${_chainId === 1 ? "etherscan" : "gnosisscan"}.io`],
+      },
+    ]);
+  } catch {
+    console.error("Failed to add network");
+  }
+}
+
+async function stressTestWalletRpc(provider: ethers.providers.Web3Provider) {
+  const success: Promise<string | boolean>[] = [];
+
+  for (let i = 0; i < 10; i++) {
+    success.push(testNonceBitmapEthCall(provider));
+  }
+
+  // if the test takes too long, we'll just assume it's not working
+  const timeoutPromise = new Promise<[false]>((resolve) => {
+    setTimeout(() => {
+      resolve([false]);
+    }, 7000);
```

Comment:
Why seven seconds?

Reply:
It's arbitrary really, it's long enough to give sufficient time but not too long before rejecting. They resolve near instantly when the RPC is good but it can take a couple seconds for them to go through sometimes as MM makes a handful of calls on load first